### DISCUSSION TITLE
Allow more modular plugin configuration

### DISCUSSION THREAD

**Discussion by jdsee** [nix-community/nixvim#1207] [General]
Hi :wave: first of all, thanks for all the work on this cool project!

When trying to port my Lua config to Nixvim I stumbled over a couple of things. Adding extra Lua Code for specific plugins bothered me the most.

In my Lua config I like to have one file per plugin and try to contain everything related in that one file. I see the following advantages in that approach:

- easy to find and edit
- code can be excluded easily when plugin is disabled or lazy loaded
- removing a plugin would simply mean to remove that one file

I tried to replicate this style with Nixvim (one module per plugin) but struggled with custom Lua code for specific plugins. I would like to avoid squeezing all the plugin specific code in `extraConfigLua` for the previously stated reasons.

Personally I would really like, if every plugin had `extraConfigLua` as a default key. That Lua code should then only be included if the plugin is enabled. The same approach would be really helpful for `extraPlugins` since those don't have any options declared.

Another fortunate side effect would be the simplification of the (potentially) upcoming lazy loading. If I'm not mistaken currently it would lead to problems if a lazily loaded plugin was `require`d from `extraConfigLua`.

So I'm wondering, if others can relate to this issue or if I'm just missing out on something. Anyone an opinion on this?

**Comment by siph** [comment]
I may have misunderstood something but you can have multiple `extraConfigLua` values declared. It doesn't help with eventual lazy-loading but you can keep the relevant parts in one file.

**Reply by jdsee** [reply]
Thanks for the tip. That gets me half way there. I’d still like to have the extra Lua code only included when the plugin is enabled. 

Probably my concern is actually more related to lazy loading (which isn’t implemented yet). From my understanding it would be necessary then to separate the extra Lua code for each plugin.

I’ll experiment a little more with the global `extraConfigLua`. Thanks again ✌️

**Reply by traxys** [reply]
You could use lib.mkIf config.plugins.xxx.enable to do that!

**Reply by jdsee** [reply]
Yea, but that wouldn’t exclude the extra Lua config from the generated Lua config. It’s not a big deal in the end. The more I think about it, the more I realise that lazy loading is probably what I’m really aiming at and that’s already discussed in other issues. Thanks for the help :)

**Reply by traxys** [reply]
You would be able to:

```nix
{config, lib, ...}: {
   plugins.foobar = {
       enable = someCondition;
       settings = { .... };
   };

   extraConfigLua = lib.mkIf config.plugins.foobar.enable ''
       --  lua code
   '';
}
```

This will add the extra lua code to the lua config only if the plugin is enabled

**Reply by jdsee** [reply]
Ohh I didn’t get it at first. This is cool. Thanks for the example!

**Reply by traxys** [reply]
Yeah `NixOS` modules are really powerful! You can do lot's of quite advanced stuff with them (this is how nixvim is implemented too!).

**Reply by jdsee** [reply]
It’s really great. Most of my setup is configured with Nix but I’m still getting confused sometimes. Thanks for shedding some light on the mkIf :)

I’m really excited about Nixvim. Just missing the lazy loading (and with that the ability to declare dependencies between the modules).

Another (slightly unrelated) thing that confused me was the incoherent definition of keymaps in different plugins. Would it be helpful if I wrote an issue for unifying the keymap schemas between all plugins?

**Reply by traxys** [reply]
Dependencies should be managed automatically by nixpkgs itself (and for some plugins nixvim).

Unifying keymaps (introduced by nixvim directly, not a plugin option) is a future goal yes.

**Reply by alexandru0-dev** [reply]
@traxys is there a way to get the `config` argument as a recursive reference when using in combination with `makeNixvimWithModule`? (so flake with nixvim config)

**Reply by MattSturgeon** [reply]
Not sure what you mean by "recursive reference", but you can have access to the evaluated/merged config both within the nixvim modules (as a module arg) and on the resulting derivation (as an additional attribute).

```nix
let
  myNvim = makeNixvim (
    { config, options, lib, pkgs, ... }: {
      # Can access `config` as a module arg here
    }
  );

  # Or as an attr on the derivation:
  myConfig = myNixvim.config;
in
```

**Comment by GaetanLepage** [comment]
@jdsee has your question found a satisfying answer ?

**Reply by jdsee** [reply]
Thanks for catching up. I moved away from nixvim in favor of a simple lua config. I think [this comment](https://github.com/nix-community/nixvim/discussions/1207#discussioncomment-8698771) is probably the closest thing to what I was looking for at that time.

**Reply by khaneliman** [reply]
Now we have ‘plugins.<name >.luaConfig.<pre|post|content>’ which is more ergonomic
